nomad/state/state_store.go

```diff
@@ -4100,6 +4107,67 @@ type StateSnapshot struct {
 	StateStore
 }
 
+// DenormalizeAllocationsMap takes in a map of nodes to allocations, and queries the
+// Allocation for each of the Allocation diffs and merges the updated attributes with
+// the existing Allocation, and attaches the Job provided
+func (s *StateSnapshot) DenormalizeAllocationsMap(nodeAllocations map[string][]*structs.Allocation, job *structs.Job) error {
+	for _, allocDiffs := range nodeAllocations {
+		if err := s.DenormalizeAllocationsSlice(allocDiffs, job); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// DenormalizeAllocationsSlice queries the Allocation for each of the Allocation diffs and merges
+// the updated attributes with the existing Allocation, and attaches the Job provided
+func (s *StateSnapshot) DenormalizeAllocationsSlice(allocDiffs []*structs.Allocation, job *structs.Job) error {
+	// Output index for denormalized Allocations
+	j := 0
+
+	for _, allocDiff := range allocDiffs {
+		alloc, err := s.AllocByID(nil, allocDiff.ID)
+		if err != nil {
+			return fmt.Errorf("alloc lookup failed: %v", err)
+		}
+		if alloc == nil {
+			continue
```

Comment:
This should return an error that bubbles up such that the plan apply fails and the worker is forced to do a index refresh. 
Otherwise if there's a race between a forced garbage collection and the scheduler making an update to the alloc, the alloc could be gone from the state store before it gets here and silently return true though the update didn't actually make it. 
cc @dadgar to double check the above^

Reply:
I think it is safer to return an error here as it indicates that the scheduler made an update on stale information. I don't think it is likely to ever be hit though because it not being in the state store means that the allocation has been GC'd which is only possible if the user forced GC'd between the scheduler snapshot and plan apply.